crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -391,16 +393,40 @@ impl<'db> Class<'db> {
                     );
                 }
                 ClassBase::Class(class) => {
-                    if let member @ SymbolAndQualifiers(Symbol::Type(_, _), _) =
+                    if let member @ SymbolAndQualifiers(Symbol::Type(ty, boundness), _) =
                         class.own_instance_member(db, name)
                     {
-                        return member;
+                        if boundness == Boundness::Bound {
+                            if union.is_empty() {
+                                // Short-circuit, no need to allocate inside the union builder
+                                return member;
+                            }
+
+                            return SymbolAndQualifiers(
+                                Symbol::bound(union.add(ty).build()),
+                                TypeQualifiers::empty(),
+                            );
+                        }
+
+                        // If we see a possibly-unbound symbol, we need to keep looking
+                        // higher up in the MRO.
+                        union = union.add(ty);
                     }
                 }
             }
         }
 
-        SymbolAndQualifiers(Symbol::Unbound, TypeQualifiers::empty())
+        if union.is_empty() {
+            SymbolAndQualifiers(Symbol::Unbound, TypeQualifiers::empty())
+        } else {
+            // If we have reached this point, we know that we have only seen possibly-unbound symbols.
+            // This means that the final result is still possibly-unbound.
+
+            SymbolAndQualifiers(
+                Symbol::Type(union.build(), Boundness::PossiblyUnbound),
+                TypeQualifiers::empty(),
```

Comment:
Here also we are throwing away type-qualifier information from whatever superclass attribute(s) we found.